File: platforms/flutter/colyseus/lib/src/input_handle.dart
import 'dart:ffi';

import 'bindings/colyseus_core.dart';
import 'colyseus.dart';
import 'schema_view.dart';

/// How an input channel reaches the server.
enum InputMode {
  /// Ordered and guaranteed. Required for anything the client predicts,
  /// because reconciliation replays the buffered inputs the server acked.
  reliable,

  /// Fire-and-forget. Rejected for now — it needs a WebTransport datagram
  /// channel, which this SDK does not have. [ColyseusRoom.input] throws.
  unreliable,
}

/// Configuration for [ColyseusRoom.input]. First call wins — the handle is
/// created lazily and later calls return the existing one.
class InputOptions {
  /// Delivery mode; defaults to [InputMode.reliable].
  final InputMode mode;

  /// How many seconds of inputs to keep for replay. 0 uses the default (3).
  final int historySize;

  /// Lag-compensation render delay in milliseconds.
  ///
  /// Leave at 0 and the predict layer binds it from its own interpolation
  /// delay — which is what you want, since the server rewinds to the instant
  /// you actually displayed. Set it by hand only when interpolating outside
  /// the predict layer.
  final double renderDelay;

  /// Decides per send whether this input carries a lag-compensation stamp.
  ///
  /// Stamping lets the server rewind other entities to what this client saw.
  /// Gate it on the inputs that need it (`(data) => data.getBool('fire')`)
  /// rather than stamping everything.
  final bool Function(SchemaView data)? allowRewind;

  const InputOptions({
    this.mode = InputMode.reliable,
    this.historySize = 0,
    this.renderDelay = 0,
    this.allowRewind,
  });
}

/// A typed input channel to the server.
///
/// Inputs are not messages: they ride a dedicated channel with sequence
/// numbers and acknowledgements, which is what makes prediction and
/// reconciliation possible. The server's ack rides back on the same channel
/// and drives the room clock, so a predicting client sends one input per tick
/// even when nothing changed.
///
/// Mutate [data], then [send]:
///
/// ```dart
/// final input = room.input();
/// for (var i = 0; i < steps; i++) {
///   input.data['moveX'] = keyboard.moveX;
///   input.data['moveY'] = keyboard.moveY;
///   input.send();
/// }
/// ```
///
/// The schema comes from the server's handshake, so field names match the
/// room's `defineInput()` declaration.
class InputHandle {
  final Pointer<colyseus_input_handle> _handle;
  SchemaView? _data;

  InputHandle(this._handle);

  /// The native handle.
  Pointer<colyseus_input_handle> get handle => _handle;

  /// The mutable input payload.
  ///
  /// One instance reused across sends — set the fields that changed and the
  /// encoder deltas the rest.
  SchemaView get data =>
      _data ??= SchemaView(core.colyseus_input_handle_data(_handle).address);

  /// Transmits the current [data] and returns its sequence number.
  ///
  /// Returns 0 when nothing was sent. Sequences start at 1 and are what
  /// reconcilers key their replay buffer on.
  int send() => core.colyseus_input_handle_send(_handle);

  /// Clears round-trip state and starts a new epoch.
  ///
  /// The next send emits a full snapshot instead of a delta. Reconcilers watch
  /// the epoch and self-reset, which is why this is the right call after a
  /// reconnect — the server counts sequences from zero again.
  void reset() => core.colyseus_input_handle_reset(_handle);

  /// How many inputs have been sent since the last [reset].
  int get sentCount => core.colyseus_input_handle_sent_count(_handle);

  /// The highest sequence the server has confirmed processing.
  int get lastProcessed => core.colyseus_input_handle_last_processed(_handle);

  /// Inputs sent but not yet acknowledged — the depth of the replay buffer,
  /// and a direct read on how much the client is predicting ahead.
  int get pendingCount => core.colyseus_input_handle_pending_count(_handle);

  /// Increments on every [reset]; observers compare it to detect one.
  int get epoch => core.colyseus_input_handle_epoch(_handle);

  /// Server tick rate in Hz, or 0 when the server didn't advertise one.
  int get tickRate => core.colyseus_input_handle_tick_rate(_handle);

  /// Server patch rate in Hz, or 0 when not advertised.
  int get patchRate => core.colyseus_input_handle_patch_rate(_handle);

  /// Simulation sub-steps per tick; at least 1.
  int get subSteps => core.colyseus_input_handle_sub_steps(_handle);

  /// Lag-compensation render delay in milliseconds.
  double get renderDelay => core.colyseus_input_handle_render_delay(_handle);

  set renderDelay(double milliseconds) =>
      core.colyseus_input_handle_set_render_delay(_handle, milliseconds);
}

File: platforms/swift/Sources/Colyseus/InputHandle.swift
import CColyseus
import Foundation

public extension Colyseus {
    /// The client's input channel for a room that declares `defineInput()`.
    ///
    /// One input is consumed per server tick. Write the fields, call
    /// ``send()``, and the sequence number that comes back is what the
    /// reconciler and the lag-compensation window key off.
    ///
    /// ```swift
    /// let input = room.input()
    /// for _ in 0 ..< predict.tick(room.clock.now) {
    ///     input.data.set("moveX", to: keyboard.x)
    ///     input.send()
    /// }
    /// ```
    ///
    /// The schema comes from the handshake, so nothing has to be generated for
    /// it.
    struct InputHandle: @unchecked Sendable {
        let raw: OpaquePointer

        /// The instance to write this tick's input into. Reused every tick —
        /// whatever you leave in it is what gets sent.
        public var data: SchemaView {
            SchemaView(colyseus_input_handle_data(raw))!
        }

        /// Transmit the current values. Returns the sequence number, or 0 when
        /// the room is not open.
        ///
        /// Always sends while connected, even when nothing changed: the server
        /// consumes exactly one input per tick, and a gap is a stutter.
        @discardableResult
        public func send() -> Int {
            Int(colyseus_input_handle_send(raw))
        }

        /// Start over: the next send is a full snapshot and the sequence
        /// numbering restarts.
        ///
        /// A reconnected session counts inputs from zero, so this belongs in
        /// ``Colyseus/Room/onReconnect(_:)`` alongside resetting whatever
        /// predicts from it.
        public func reset() {
            colyseus_input_handle_reset(raw)
        }

        /// The input that was sent as `seq`, for replaying it.
        ///
        /// The instance behind this is REUSED across calls — read what you
        /// need before asking for another.
        public func at(_ seq: Int) -> SchemaView? {
            SchemaView(colyseus_input_handle_at(raw, Int32(seq)))
        }

        /// When the client rendered the world it stamped onto input `seq` —
        /// the instant a lag-compensated server rewinds to.
        public func reckonTime(at seq: Int) -> Double {
            colyseus_input_handle_reckon_time_at(raw, Int32(seq))
        }

        /// How far behind live the app draws remote entities.
        ///
        /// This must match the interpolation delay actually used, or a
        /// rewinding server rewinds to the wrong instant. A reconciler built
        /// from a ``Colyseus/Predict`` sets it for you.
        public var renderDelay: Double {
            get { colyseus_input_handle_render_delay(raw) }
            nonmutating set { colyseus_input_handle_set_render_delay(raw, newValue) }
        }

        /// Which inputs the server should rewind for — a fire button, say.
        /// Rewinding every input would be expensive and pointless.
        public func allowRewind(_ predicate: @escaping @Sendable (SchemaView) -> Bool) {
            let box = RewindGate { instance in
                guard let view = SchemaView(instance) else { return false }
                return predicate(view)
            }
            // Held for the handle's lifetime: the core calls it per input and
            // there is no unregister.
            RewindGate.keepAlive(box)

            colyseus_input_handle_set_allow_rewind(raw, { instance, userdata in
                borrowObject(userdata, as: RewindGate.self)?.body(instance) ?? false
            }, retainedPointer(box))
        }

        // MARK: - Round-trip state

        /// How many inputs have been sent.
        public var sentCount: Int { Int(colyseus_input_handle_sent_count(raw)) }

        /// The last sequence the server confirmed.
        public var lastProcessed: Int { Int(colyseus_input_handle_last_processed(raw)) }

        /// Inputs applied locally that the server has not confirmed. At 200 ms
        /// on a 20 Hz room this sits around three or four.
        public var pendingCount: Int { Int(colyseus_input_handle_pending_count(raw)) }

        /// How many past inputs are kept for replay.
        public var replayBufferSize: Int { Int(colyseus_input_handle_replay_buffer_size(raw)) }

        /// Bumped by ``reset()``. Anything holding sequence numbers across a
        /// reconnect should notice this changing.
        public var epoch: Int { Int(colyseus_input_handle_epoch(raw)) }

        // MARK: - What the server advertised

        /// Server ticks per second.
        public var tickRate: Int { Int(colyseus_input_handle_tick_rate(raw)) }

        /// State patches per second.
        public var patchRate: Int { Int(colyseus_input_handle_patch_rate(raw)) }

        /// Simulation sub-steps the server runs per tick.
        public var subSteps: Int { Int(colyseus_input_handle_sub_steps(raw)) }
    }

    /// How an input channel behaves.
    struct InputOptions: Sendable {
        /// How many past inputs to keep for replay, in multiples of the
        /// round trip.
        public var historySize: Int = 3

        /// The interpolation delay the app draws remote entities at.
        public var renderDelay: Double = 0

        public init(historySize: Int = 3, renderDelay: Double = 0) {
            self.historySize = historySize
            self.renderDelay = renderDelay
        }
    }
}

public extension Colyseus.Room {
    /// The room's input channel, built on first use.
    ///
    /// Returns nil for a room whose server never calls `defineInput()` — there
    /// is no input schema to build one from.
    func input(_ options: Colyseus.InputOptions = .init()) -> Colyseus.InputHandle? {
        var raw = colyseus_input_options_t()
        raw.unreliable = false
        raw.history_size = Int32(options.historySize)
        raw.render_delay = options.renderDelay

        // A NULL vtable means "build it from the handshake reflection", which
        // is the only path a binding without codegen has.
        guard let handle = colyseus_room_input(self.raw, nil, &raw) else { return nil }
        return Colyseus.InputHandle(raw: handle)
    }
}

/// The rewind predicate is called by the core per input and never
/// unregistered, so it outlives every Swift reference to it.
private final class RewindGate: @unchecked Sendable {
    let body: (UnsafeMutableRawPointer?) -> Bool

    init(_ body: @escaping (UnsafeMutableRawPointer?) -> Bool) {
        self.body = body
    }

    private static let retained = Guarded<[RewindGate]>([])

    static func keepAlive(_ gate: RewindGate) {
        retained.withLock { $0.append(gate) }
    }
}
